parsl/executors/high_throughput/mpi_resource_management.py

```diff
@@ -0,0 +1,222 @@
+    2) put_task for execution onto workers:
+        a) if resources are available attach resource list
+        b) if unavailable place tasks into backlog
+    3) get_result will fetch a result and relinquish nodes,
+       and attempt to schedule tasks in backlog if any.
+    """
+    def __init__(
+        self,
+        pending_task_q: multiprocessing.Queue,
+        pending_result_q: multiprocessing.Queue,
+    ):
+        super().__init__(pending_task_q, pending_result_q)
+        self.scheduler = identify_scheduler()
+        # PriorityQueue is threadsafe
+        self._backlog_queue: queue.PriorityQueue = queue.PriorityQueue()
+        self._map_tasks_to_nodes: Dict[str, List[str]] = {}
+        self.available_nodes = get_nodes_in_batchjob(self.scheduler)
+        self._free_node_counter = SpawnContext.Value("i", len(self.available_nodes))
+        # mp.Value has issues with mypy
+        # issue https://github.com/python/typeshed/issues/8799
+        # from mypy 0.981 onwards
+        self.nodes_q: queue.Queue = queue.Queue()
+        for node in self.available_nodes:
+            self.nodes_q.put(node)
+
+        logger.info(
+            f"Starting MPITaskScheduler with {len(self.available_nodes)}"
+        )
+
+    def _get_nodes(self, num_nodes: int) -> List[str]:
+        """Thread safe method to acquire num_nodes from free resources
+
+        Raises: MPIResourceUnavailable if there aren't enough resources
+        Returns: List of nodenames:str
+        """
+        logger.debug(
+            f"Requesting : {num_nodes=} we have {self._free_node_counter}"
+        )
+        acquired_nodes = []
+        with self._free_node_counter.get_lock():
+            if num_nodes <= self._free_node_counter.value:  # type: ignore[attr-defined]
+                self._free_node_counter.value -= num_nodes  # type: ignore[attr-defined]
+            else:
+                raise MPIResourceUnavailable(
+                    requested=num_nodes, available=self._free_node_counter.value  # type: ignore[attr-defined]
+                )
+
+            for i in range(num_nodes):
+                node = self.nodes_q.get()
+                acquired_nodes.append(node)
+        return acquired_nodes
+
+    def _return_nodes(self, nodes: List[str]) -> None:
+        """Threadsafe method to return a list of nodes"""
+        for node in nodes:
+            self.nodes_q.put(node)
+        with self._free_node_counter.get_lock():
+            self._free_node_counter.value += len(nodes)  # type: ignore[attr-defined]
+
+    def put_task(self, task_package: dict):
+        """Schedule task if resources are available otherwise backlog the task"""
+        user_ns = locals()
+        user_ns.update({"__builtins__": __builtins__})
+        _f, _args, _kwargs, resource_spec = unpack_res_spec_apply_message(
+            task_package["buffer"], user_ns, copy=False
+        )
+
+        nodes_needed = resource_spec.get("NUM_NODES")
+        if nodes_needed:
```

Comment:
whats the behaviour if you don't request any nodes with NUM_NODES? no nodes get allocated and the task gets to run anyway? and hopefully doesn't do a $PARSL_MPI_PREFIX invocation?

Reply:
No nodes get allocated, task gets to run, but you still get a `$PARSL_MPI_PREFIX`.  I'd assume that it is a user-bug if they put in `$PARSL_MPI_PREFIX` in their app and miss the resource_specification, and you'd get a broken invocation that will error out.

Reply:
so i guess you get to run up to maximum workers worth of tasks like this and then they queue like any other task?